Un-break builds & document unwrapping behaviour in to_pandas()

Commit 28f6be45733e5cb7549c4c6dae215a28b6a0e676 broke the Travis build: 
```
=================================== FAILURES ===================================
___________________ TestGenericFlatDataSet.test_write2pandas ___________________
self = <tests.test_dataset.TestGenericFlatDataSet object at 0x7ffb888ca470>
msg = <DataMessage header=<Header error=None id='Generic' prepared='2010-01-04T16:21:49+01:00' receiver=N… footer=None respo...taflow=<DataflowDefinition: 'STR1'=''> observation_dimension=<pandasdmx.model._AllDimensions object at 0x7ffb89e916a0>>
    def test_write2pandas(self, msg):
        data_series = pandasdmx.to_pandas(msg.data[0])
        assert isinstance(data_series, pd.Series)
        data_series2 = pandasdmx.to_pandas(msg.data)
        assert isinstance(data_series2, pd.Series)
>       assert data_series == data_series2
tests/test_dataset.py:58: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = FREQ  CURRENCY  CURRENCY_DENOM  EXR_TYPE  EXR_VAR  TIME_PERIOD
M     CHF       EUR             SP00      E        2010...    2010-09        True
                                                   2010-10        True
Name: value, dtype: bool
    def __nonzero__(self):
        raise ValueError(
            "The truth value of a {0} is ambiguous. "
            "Use a.empty, a.bool(), a.item(), a.any() or a.all().".format(
>               self.__class__.__name__
            )
        )
E       ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pandas/core/generic.py:1556: ValueError
```

At the same time, the documentation of `to_pandas()` should be expanded to describe this behaviour.

--- a/doc/api.rst
+++ b/doc/api.rst
@@ -45,27 +45,34 @@ SDMX-JSON
 
 ``writer``: Convert SDMX to pandas objects
 ------------------------------------------
-Convert :mod:`model <pandasdmx.model>` and :mod:`message <pandasdmx.message>`
-objects to :mod:`pandas` objects.
-
-.. versionchanged:: 0.6
-
-   Two writers are available:
-
-   - ``data2pandas`` exports a dataset or portions thereof to a pandas DataFrame
-     or Series.
-   - ``structure2pd`` exports structural metadata such as lists of data-flow
-     definitions, code-lists, concept-schemes etc. which are contained in a
-     structural SDMX message as a dict mapping resource names (e.g. 'dataflow',
-     'codelist') to pandas DataFrames.
-
 .. versionchanged:: 1.0
 
-   The :meth:`write <pandasdmx.writer.write>` method
-   (:meth:`pandasdmx.to_pandas`) handles all types of objects.
+   :meth:`pandasdmx.to_pandas` (via :meth:`write <pandasdmx.writer.write>`)
+   handles all types of objects, replacing the earlier, separate
+   ``data2pandas`` and ``structure2pd`` writers.
 
 .. automodule:: pandasdmx.writer
    :members:
+   :exclude-members: write
+
+   .. automethod:: pandasdmx.writer.write
+
+      .. autosummary::
+         write_component
+         write_dataset
+         write_dict
+         write_dimensiondescriptor
+         write_itemscheme
+         write_list
+         write_nameableartefact
+         write_response
+         write_serieskeys
+         write_structuremessage
+
+.. todo::
+   Support selection of language for conversion of
+   :class:`InternationalString <pandasdmx.model.InternationalString>`.
+
 
 
 ``remote``: Access SDMX REST web services

--- a/doc/example.rst
+++ b/doc/example.rst
@@ -38,7 +38,7 @@ to obtain data on Greece, Ireland and Spain only.
         )
 
     # Convert to a pandas.Series and select on the 'AGE' dimension
-    data = (sdmx.to_pandas(resp.data[0])
+    data = (sdmx.to_pandas(resp)
                 .xs('TOTAL', level='AGE', drop_level=False))
 
     # Explore the data set. First, show dimension names

--- a/pandasdmx/model.py
+++ b/pandasdmx/model.py
@@ -86,6 +86,9 @@ class InternationalString:
         # Using a bare string, implicitly for the DEFAULT_LOCALE
         f.name = "Name in DEFAULT_LOCALE language"
 
+    Only the first method preserves existing localizations; the latter three
+    replace them.
+
     """
     localizations: Dict[str, str] = {}
 
@@ -98,6 +101,8 @@ class InternationalString:
             value = {value[0]: value[1]}
         elif value is None:
             value = dict(kwargs)
+        elif isinstance(value, dict):
+            pass
         else:
             raise ValueError(value, kwargs)
 
@@ -148,7 +153,7 @@ class InternationalString:
         yield cls.__validate
 
     @classmethod
-    def __validate(cls, value, values, field):
+    def __validate(cls, value, values, config, field):
         if not isinstance(value, InternationalString):
             value = InternationalString(value)
 

--- a/pandasdmx/util.py
+++ b/pandasdmx/util.py
@@ -162,7 +162,6 @@ class DictLike(OrderedDict[KT, VT]):
             return result
 
 
-
 def validate_dictlike(*fields):
     def decorator(cls):
         v = make_generic_validator(DictLike.validate)

--- a/pandasdmx/writer.py
+++ b/pandasdmx/writer.py
@@ -44,10 +44,6 @@ def write(obj, *args, **kwargs):
     converted using :meth:`write_dataset`. See individual ``write_*`` methods
     named for more information on their behaviour, including accepted *args*
     and *kwargs*.
-
-    .. todo::
-       Support selection of language for conversion of
-       :class:`InternationalString <pandasdmx.model.InternationalString>`.
     """
     cls = obj.__class__
     function = 'write_' + _alias.get(cls, cls).__name__.lower()
@@ -56,11 +52,23 @@ def write(obj, *args, **kwargs):
 
 # Functions for Python containers
 def write_list(obj, *args, **kwargs):
-    """List of objects."""
+    """Convert a :class:`list` of SDMX objects.
+
+    For the following *obj*, :meth:`write_list` returns :class:`pandas.Series`
+    instead of a :class:`list`:
+
+    - a list of :class:`Observation <pandasdmx.model.Observation>`:
+      the Observations are written using :meth:`write_dataset`.
+    - a list with only 1 :class:`DataSet <pandasdmx.model.DataSet>` (e.g. the
+       :attr:`data <pandasdmx.message.DataMessage.data>` attribute of
+       :class:`DataMessage <pandasdmx.message.DataMessage>`): the Series for
+       the single element is returned.
+    - a list of :class:`SeriesKey`: the key values (but no data) are returned.
+    """
     if isinstance(obj[0], Observation):
         return write_dataset(obj, *args, **kwargs)
     elif isinstance(obj[0], DataSet) and len(obj) == 1:
-        return write_dataset(obj[0], *args, **kwargs)
+        return write(obj[0], *args, **kwargs)
     elif isinstance(obj[0], SeriesKey):
         return write_serieskeys(obj, *args, **kwargs)
     else:
@@ -98,11 +106,16 @@ def write_dict(obj, *args, **kwargs):
 
 # Functions for message classes
 def write_response(obj, *args, **kwargs):
-    """Writing a pandasdmx.api.Response → write the Response.msg."""
+    """Convert :class:`pandasdmx.api.Response`.
+
+    The :attr:`msg <pandasdmx.api.Response.msg>` attribute of *obj* is
+    converted.
+    """
     return write(obj.msg, *args, **kwargs)
 
 
 def write_datamessage(obj, *args, **kwargs):
+    """Convert :class:`DataMessage <pandasdmx.message.DataMessage>`."""
     if len(obj.data) == 1:
         return write(obj.data[0], *args, **kwargs)
     else:
@@ -110,7 +123,7 @@ def write_datamessage(obj, *args, **kwargs):
 
 
 def write_structuremessage(obj, include=None, **kwargs):
-    """Transform :class:`pandasdmx.message.StructureMessage` to pandas.
+    """Convert :class:`StructureMessage <pandasdmx.message.StructureMessage>`.
 
     Parameters
     ----------
@@ -154,12 +167,18 @@ def write_structuremessage(obj, include=None, **kwargs):
 # Functions for model classes
 
 def write_component(obj):
+    """Convert :class:`Component <pandasdmx.model.Component>`.
+
+    The :attr:`Concept.id <pandasdmx.model.Concept.id>` attribute of the
+    :attr:`Component.concept_identity
+    <pandasdmx.model.Component.concept_identity>` is returned.
+    """
     return str(obj.concept_identity.id)
 
 
 def write_dataset(obj, attributes='', dtype=np.float64, constraint=None,
                   fromfreq=False, parse_time=True):
-    """Write :class:`DataSet <pandasdmx.model.DataSet>`.
+    """Convert :class:`DataSet <pandasdmx.model.DataSet>`.
 
     Parameters
     ----------
@@ -246,13 +265,23 @@ def write_dataset(obj, attributes='', dtype=np.float64, constraint=None,
 
 
 def write_dimensiondescriptor(obj):
+    """Convert :class:`DimensionDescriptor
+    <pandasdmx.model.DimensionDescriptor>`.
+
+    The :attr:`components <pandasdmx.model.DimensionDescriptor.components>` of
+    the DimensionDescriptor are written.
+    """
     return write(obj.components)
 
 
 def write_itemscheme(obj, locale=DEFAULT_LOCALE):
-    """Convert a model.ItemScheme object to a pd.Series.
+    """Convert :class:`ItemScheme <pandasdmx.model.ItemScheme>`.
 
     Names from *locale* are serialized.
+
+    Returns
+    -------
+    pandas.Series
     """
     items = {}
     seen = set()
@@ -294,10 +323,16 @@ def write_itemscheme(obj, locale=DEFAULT_LOCALE):
 
 
 def write_nameableartefact(obj):
+    """Convert :class:`NameableArtefact <pandasdmx.model.NameableArtefact>`.
+
+    The :attr:`name <pandasdmx.model.NameableArtefact.name>` attribute of *obj*
+    is returned.
+    """
     return str(obj.name)
 
 
 def write_serieskeys(obj):
+    """Convert a list of :class:`SeriesKey <pandasdmx.model.SeriesKey>`."""
     result = []
     for sk in obj:
         result.append({dim: kv.value for dim, kv in sk.order().values.items()})

--- a/pandasdmx/writer.py
+++ b/pandasdmx/writer.py
@@ -59,7 +59,7 @@ def write_list(obj, *args, **kwargs):
     """List of objects."""
     if isinstance(obj[0], Observation):
         return write_dataset(obj, *args, **kwargs)
-    elif isinstance(obj[0], DataSet):
+    elif isinstance(obj[0], DataSet) and len(obj) == 1:
         return write_dataset(obj[0], *args, **kwargs)
     elif isinstance(obj[0], SeriesKey):
         return write_serieskeys(obj, *args, **kwargs)

--- a/setup.py
+++ b/setup.py
@@ -1,11 +1,10 @@
 from setuptools import find_packages, setup
-from codecs import open
 
 
 INSTALL_REQUIRES = [
     'lxml>=3.6',
     'pandas>=0.20',
-    'pydantic>=0.25',
+    'pydantic>=0.31',
     'requests>=2.7',
     'setuptools>19',
     ]

--- a/tests/test_dataset.py
+++ b/tests/test_dataset.py
@@ -1,4 +1,5 @@
 import pandas as pd
+import pandas.testing as pdt
 import pandasdmx
 from pandasdmx import message, model
 
@@ -50,12 +51,16 @@ class TestGenericFlatDataSet(DataMessageTest):
         assert o0.attrib.OBS_STATUS == 'A'
         assert o0.attrib.DECIMALS == '4'
 
-    def test_write2pandas(self, msg):
+    def test_to_pandas(self, msg):
+        # Single data series is converted to pd.Series
         data_series = pandasdmx.to_pandas(msg.data[0])
         assert isinstance(data_series, pd.Series)
-        data_series2 = pandasdmx.to_pandas(msg.data)
-        assert isinstance(data_series2, pd.Series)
-        assert data_series == data_series2
+
+        # When len(msg.data) is 1, the data series in a single Dataset are
+        # unwrapped automatically
+        assert len(msg.data) == 1
+        data_series2 = pandasdmx.to_pandas(msg.data)  # NB no '[0]' index
+        pdt.assert_series_equal(data_series, data_series2)
 
 
 class TestGenericSeriesDataSet(DataMessageTest):

--- a/tests/test_model.py
+++ b/tests/test_model.py
@@ -70,14 +70,24 @@ def test_internationalstring():
     assert len(i.name.localizations) == 2
     assert i.name['FR'] == 'Banque centrale européenne'
 
+    # repr() gives all localizations
+    assert repr(i.name) == '\n'.join(sorted([
+        'DE: Europäische Zentralbank',
+        'FR: Banque centrale européenne',
+        ]))
+
     # Setting with a string directly sets the value in the default locale
     i.name = 'European Central Bank'
-    assert len(i.name.localizations) == 3
+    assert len(i.name.localizations) == 1
     assert i.name.localizations[DEFAULT_LOCALE] == 'European Central Bank'
 
     # Setting with a (locale, text) tuple
     i.name = ('FI', 'Euroopan keskuspankki')
-    assert len(i.name.localizations) == 4
+    assert len(i.name.localizations) == 1
+
+    # Setting with a dict()
+    i.name = {'IT': 'Banca centrale europea'}
+    assert len(i.name.localizations) == 1
 
     # Using some other type is an error
     with raises(pydantic.ValidationError):
@@ -88,13 +98,6 @@ def test_internationalstring():
 
     # str() uses the default locale
     assert str(i2.name) == 'European Central Bank'
-
-    # repr() gives all localizations
-    assert repr(i.name) == '\n'.join(sorted([
-        '{}: European Central Bank'.format(DEFAULT_LOCALE),
-        'DE: Europäische Zentralbank',
-        'FR: Banque centrale européenne',
-        'FI: Euroopan keskuspankki']))
 
 
 def test_item():

